model/PermissionsService.php

```diff
@@ -54,61 +52,162 @@ public function __construct(
-
-    private function saveResourcePermissions(
-        core_kernel_classes_Resource $resource,
-        array $privilegesToSet,
-        bool $isRecursive
-    ): void {
-        $currentPrivileges = $this->dataBaseAccess->getResourcePermissions($resource->getUri());
-        $addRemove = $this->strategy->normalizeRequest($currentPrivileges, $privilegesToSet);
+    /**
+     * Updates the permissions for a set of resources based on the ACLs, root
+     * resource and recursion parameters contained in the provided command.
+     *
+     * - For recursive commands having a class as the root, updates permissions
+     *   for that class and all its descendants (i.e. updates all resources AND
+     *   classes using the provided root class as the initial node for a tree
+     *   traversal, which may be slow and resource-intensive). This is provided
+     *   for backward compatibility purposes.
+     *
+     * - For non-recursive commands having a class as the root AND including
+     *   nested resources, updates permissions for the class and all instances
+     *   of that class, but skips all nested classes and instances of them (i.e.
+     *   does not go down into nested levels of the resource tree).
+     *
+     * - Otherwise (i.e. non-class roots), it updates only the resource set as
+     *   the root resource for the command.
+     */
+    public function applyPermissions(ChangePermissionsCommand $command): void
+    {
+        $resources = $this->getResourcesToUpdate($command);
+        $currentPermissions = $this->getResourcesPermissions($resources);
+        $permissionsDelta = $this->getDeltaForResourceTree($command, $currentPermissions);
 
-        if (empty($addRemove)) {
+        if (empty($permissionsDelta)) {
             return;
         }
 
-        $resourcesToUpdate = $this->getResourcesToUpdate($resource, $isRecursive);
-        $permissionsList = $this->getResourcesPermissions($resourcesToUpdate);
-        $actions = $this->getActions($resourcesToUpdate, $permissionsList, $addRemove);
-
-        $this->dryRun($actions, $permissionsList);
+        $actions = $this->getActions($resources, $currentPermissions, $permissionsDelta);
+        $this->dryRun($actions, $currentPermissions);
         $this->wetRun($actions);
-        $this->triggerEvents($addRemove, $resource->getUri(), $isRecursive);
+
+        $this->triggerEvents(
+            $permissionsDelta[$command->getRoot()->getUri()],
+            $command->getRoot()->getUri(),
+            $command->applyToNestedResources()
+        );
+    }
+
+    /**
+     * @deprecated Use applyPermissions() instead
+     * @see applyPermissions
+     */
+    public function saveResourcePermissionsRecursive(
+        core_kernel_classes_Resource $resource,
+        array $privilegesToSet
+    ): void {
+        $this->applyPermissions(
+            (new ChangePermissionsCommand($resource, $privilegesToSet))
+                ->withRecursion()
+                ->withNestedResources()
```

Comment:
Default is already false :) 

```suggestion
            (new ChangePermissionsCommand($resource, $privilegesToSet))
```

Reply:
Actually, here I'm overriding the default `false` values from the command and setting them to `true` instead on purpose (so this method keeps its old behaviour).

- `withRecursion()` sets `ChangePermissionsCommand::$isRecursive=true`
- `withNestedResources()` sets `ChangePermissionsCommand::$applyToNestedResources=true`

Both of them are `false` by default when `ChangePermissionsCommand` has just been created.

Reply:
True. Misread :)